Add resource_compute_region_autoscaler

usererror opened the pull request:
# Why
Attempts to fix https://github.com/terraform-providers/terraform-provider-google/issues/535.

It seemed easy enough based on @rosbo 's comment and I think this works but please let me know if I've missed anything! I tried to copy the regular resource_compute_autoscaler as much as possible and luckily they both use the same underlying type.

# Tests
I ran acceptance tests pointing at one of my google projects and they all passed.

```shell
➜  google git:(region-autoscaler) ✗ TF_ACC=1 go test -v -run TestAccComputeRegionAutoscaler
=== RUN   TestAccComputeRegionAutoscaler_basic
--- PASS: TestAccComputeRegionAutoscaler_basic (70.90s)
=== RUN   TestAccComputeRegionAutoscaler_update
--- PASS: TestAccComputeRegionAutoscaler_update (83.31s)
PASS
ok      github.com/terraform-providers/terraform-provider-google/google    154.249s
```

usererror commented on google/resource_compute_region_autoscaler.go:
```diff
@@ -0,0 +1,372 @@
+	return resourceComputeRegionAutoscalerRead(d, meta)
+}
+
+func flattenRegionAutoscalingPolicy(policy *compute.AutoscalingPolicy) []map[string]interface{} {
```
I'm not sure this needs to be redefined for the regional autoscaler since they both use compute.Autoscaler... anyway. I didn't touch this function actually, just renamed it.

rosbo replied:
You can delete this method and reuse the one in the non-regional version

rosbo commented on google/resource_compute_region_autoscaler.go:
```diff
@@ -0,0 +1,372 @@
+				ForceNew: true,
+			},
+
+			"autoscaling_policy": &schema.Schema{
```
You can extract this &schema.Schema definition and reuse it for both autoscaler

rosbo commented on google/resource_compute_region_autoscaler.go:
```diff
@@ -0,0 +1,372 @@
+	}
+}
+
+func buildRegionAutoscaler(d *schema.ResourceData) (*compute.Autoscaler, error) {
```
You can remove this method and use the `buildAutoscaler`. In fact, I think you are not even using it and already using the `buildAutoscaler`.

rosbo commented on google/resource_compute_region_autoscaler.go:
```diff
@@ -0,0 +1,372 @@
+	d.Set("self_link", scaler.SelfLink)
+	d.Set("name", scaler.Name)
+	d.Set("target", scaler.Target)
+	d.Set("region", regionUrl[len(regionUrl)-1])
```
You can use the method `GetResourceNameFromSelfLink`

rosbo commented on google/resource_compute_region_autoscaler.go:
```diff
@@ -0,0 +1,372 @@
+
+			"autoscaling_policy": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
```
Add `MaxItems: 1` since we can only specify at most on `autoscaling_policy`. It is also not Optional since you must have at least one `autoscaling_policy`. Please add `Required: true`

rosbo commented on google/resource_compute_region_autoscaler.go:
```diff
@@ -0,0 +1,372 @@
+
+	aspCount := d.Get("autoscaling_policy.#").(int)
+	if aspCount != 1 {
+		return nil, fmt.Errorf("The autoscaler must have exactly one autoscaling_policy, found %d.", aspCount)
```
You can remove this line once this requirement is enforced by the schema like suggested above. The advantage of enforcing this in the schema is that it fails at `terraform plan` instead of `terraform apply`.

rosbo commented on google/resource_compute_region_autoscaler_test.go:
```diff
@@ -0,0 +1,262 @@
+package google
```
Can you also add import acceptance tests?

By convention, we put them in a separate file. See `import_compute_autoscaler_test.go"

usererror commented:
Thanks for the review Vincent! I fixed up the zonal autoscaler on your recommendation and got rid of the duplicate regional functions and both are much more readable now. I also added the import acceptance tests for the regional autoscaler and ran all the acceptance tests again. This was a pleasure to work on, very easy to test and work with!

# Acceptance Tests
```shell
➜  google git:(region-autoscaler) ✗ TF_ACC=1 go test -v -run TestAccComputeAutoscaler
=== RUN   TestAccComputeAutoscaler_importBasic
--- PASS: TestAccComputeAutoscaler_importBasic (69.92s)
=== RUN   TestAccComputeAutoscaler_basic
--- PASS: TestAccComputeAutoscaler_basic (68.78s)
=== RUN   TestAccComputeAutoscaler_update
--- PASS: TestAccComputeAutoscaler_update (81.43s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	220.148s
```

```shell
➜  google git:(region-autoscaler) ✗ TF_ACC=1 go test -v -run TestAccComputeRegionAutoscaler
=== RUN   TestAccComputeRegionAutoscaler_importBasic
--- PASS: TestAccComputeRegionAutoscaler_importBasic (70.69s)
=== RUN   TestAccComputeRegionAutoscaler_basic
--- PASS: TestAccComputeRegionAutoscaler_basic (70.94s)
=== RUN   TestAccComputeRegionAutoscaler_update
--- PASS: TestAccComputeRegionAutoscaler_update (83.86s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	225.508s
```

rosbo commented:
Awesome work! Thanks for adding support for this new resource.
